src/Draco.Compiler/Syntax/ParseTree.cs

```diff
@@ -0,0 +1,144 @@
+using Draco.Compiler.Utilities;
+
+namespace Draco.Compiler.Syntax;
+
+/// <summary>
+/// An immutable structure representing a parsed source text with information about concrete syntax.
+/// </summary>
+internal abstract record class ParseTree
+{
+    public sealed record class CompilationUnit(
+        ValueArray<Decl> Declarations) : ParseTree;
+
+    public abstract record class Decl : ParseTree
+    {
+        public sealed record class Func(
+            IToken FuncKeyword,
+            IToken Identifier,
+            IToken OpenParenToken,
+            ValueArray<(
+                (   IToken Identifier,
+                    TypeSpecifier Type
+                ) Param,
+                IToken CommaToken
+            )> Params,
```

Comment:
This looks incorrect and a bit convoluted as well. We are missing the colon token in between parameters. My suggestion: let's have that `PunctuatedList<TElement>` (which can use `ValueArray<T>` type and have a separate `FunctionParameter`. Then this construct can become `PunctuatedList<FunctionParameter>`.

Reply:
Nevermind about the semicolon, `TypeSpecifier` has it.